File: TopInterview150/InsertDeleteGetRandom/solution.py
class RandomizedSet(object):
    
    def __init__(self):
        self.lst = []
        self.idxMap = {}
    def search(self, val):
        return val in self.idxMap
    
    def insert(self, val):
        """
        :type val: int
        :rtype: bool
        """
        """if already in the map, return false"""
        if self.search(val):
            return False
        """if not in the map, add it and return true"""
        self.lst.append(val)
        self.idxMap[val] = len(self.lst) - 1
        return True      

    def remove(self, val):
        """
        :type val: int
        :rtype: bool
        """
        """if not in the list, return false"""
        if not self.search(val):
            return False
        """if in the list, remove it and return true"""
        idx = self.idxMap[val]
        self.lst[idx] = self.lst[-1]
        self.idxMap[self.lst[-1]] = idx
        self.lst.pop()
        del self.idxMap[val]
        return True
        

    def getRandom(self):
        """
        :rtype: int
        """
        return random.choice(self.lst)
        

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

"""
class RandomizedSet(object):
    
    def __init__(self):
        self.thisSet = set()
    def insert(self, val):
        ""
        :type val: int
        :rtype: bool
        ""
        if val in self.thisSet:
            return False
        self.thisSet.add(val)
        return True
        

    def remove(self, val):
        ""
        :type val: int
        :rtype: bool
        ""
        if val not in self.thisSet:
             return False
        self.thisSet.remove(val)
        return True
        
        

    def getRandom(self):
        ""
        :rtype: int
        ""
        return random.choice(list(self.thisSet))

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
"""

File: TopInterview150/InsertDeleteGetRandom/solution.cpp
#include <cstdlib>
#include <vector>
#include <unordered_map>

class RandomizedSet {
private:
    std::unordered_map<int, int> map;
    std::vector<int> elements;

public:
    RandomizedSet() {
    }
    
    bool insert(int val) {
        if (map.find(val) != map.end())//val already exists in map
        {return false;}
        map[val] = elements.size();
        elements.push_back(val);
        return true;
    }
    
    bool remove(int val) {
        if (map.find(val) == map.end()) //value doesn't exist
        {return false;}
        int lastElement = elements.back();
        int idx = map[val];
        elements[idx] = lastElement;//move the last element to the pos of the element to be removed
        map[lastElement] = idx;//update the map for latest element
        elements.pop_back();//remove last element
        map.erase(val);//remove the element from the map
        return true;
    }
    
    int getRandom() {
        if (elements.empty()) {
            throw std::out_of_range("no elements in this set");
        }
        int randomIndex = rand() % elements.size();//generate random index modulo size of map
        return elements[randomIndex];//return the element at the random index
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */